agent/xds/endpoints.go

```diff
@@ -530,36 +530,40 @@ type readyUpstreams struct {
 	upstreams        []structs.Upstream
 }
 
+// getReadyUpstreams returns a map containing the list of upstreams for each listener that is ready
 func getReadyUpstreams(cfgSnap *proxycfg.ConfigSnapshot) map[string]readyUpstreams {
 
 	ready := map[string]readyUpstreams{}
 	for _, l := range cfgSnap.APIGateway.Listeners {
-		//need to account for the state of the Listener when building the upstreams list
+		// Only include upstreams for listeners that are ready
 		if !cfgSnap.APIGateway.GatewayConfig.ListenerIsReady(l.Name) {
 			continue
 		}
+
+		// For each route bound to the listener
 		boundListener := cfgSnap.APIGateway.BoundListeners[l.Name]
-		//get route ref
 		for _, routeRef := range boundListener.Routes {
-			//get upstreams
-			upstreamMap := cfgSnap.APIGateway.Upstreams[routeRef]
-			for _, upstreams := range upstreamMap {
```

Comment:
The big change is removing this `for` loop - instead of picking the upstreams specific to the listener, we were looping through all upstreams for all listeners and adding them